src/compiler/mir/transform/function.rs

```diff
@@ -457,21 +453,42 @@ impl<'a> FuncTransformer<'a> {
         }
     }
 
-    fn unary_op(&mut self, op: UnaryOperator, right: &Expression<SemanticContext>) -> RValue {
+    fn unary_op(
+        &mut self,
+        ctx: &SemanticContext,
+        op: UnaryOperator,
+        right: &Expression<SemanticContext>,
+    ) -> Operand {
         let right = self.expression(right);
         match op {
-            UnaryOperator::Negate => self.mir.negate(right),
-            UnaryOperator::Not => self.mir.not(right),
+            UnaryOperator::Negate => {
+                let rv = self.mir.negate(right);
+                let ty = self.find_type(ctx.ty());
+                self.mir.temp_store(rv, ty, ctx.span())
+            }
```

Comment:
This is duplicated in 3 places, can it somehow be pulled out into a single location?